docs/source-2.0/java/client/configuration.rst

```diff
@@ -0,0 +1,506 @@
+===================
+Configuring Clients
+===================
+
+Generated Smithy Java clients are composed of a number of runtime-configurable components:
+
+* :ref:`Protocols <java-client-protocols>` - Defines how to serialize and deserialize a client request.
+* :ref:`Transports <java-client-transports>` - Manages connections and the sending of data on the wire.
+* :ref:`Auth Schemes <java-client-authSchemes>` - Add authentication/authorization information to a client request.
+* :ref:`Endpoint Resolvers <java-client-endpoints>` - Resolves the service endpoint to call.
+* :ref:`Context Properties <java-client-context>` - Set typed properties on the client.
+
+.. _java-client-protocols:
+
+---------
+Protocols
+---------
+
+A protocol defines how to (de)serialize and bind data into a request. For example, an HTTP+JSON protocol
+would handle the serialization of data into the HTTP message body as JSON and might bind some data to the
+HTTP message headers.
+
+Configure a protocol
+^^^^^^^^^^^^^^^^^^^^
+
+The Smithy IDL is protocol-agnostic and allows a service to support any number of protocols for transport and
+(de)serialization of data. Like the Smithy IDL, Smithy Java clients are also protocol-agnostic and allow users
+to configure a protocol at runtime.
+
+To set a protocol  at runtime, add the protocol to the client builder as follows:
+
+.. code-block:: java
+    :caption: Set protocol at runtime
+
+    var client = MyGeneratedClient.builder()
+        .protocol(new MyProtocol())
+        .build();
+
+.. admonition:: Important
+    :class: note
+
+    The input/output types of the configured protocol must be compatible with the configured transport.
+
+Set a default protocol
+^^^^^^^^^^^^^^^^^^^^^^
+
+While users may want to set a different protocol at runtime, it is typically desirable for generated clients to have a default protocol that can be used without any client configuration.
+
+To configure the client code generation plugin to set a default protocol, add the protocol’s fully qualified Smithy ID to the  protocol setting in the smithy-build.json configuration file.
+
+.. code-block:: json
+    :caption: smithy-build.json
+
+    "plugins": {
+        "java-client-codegen": {
+          // Set the default protocol for the client.
+          "protocol": "smithy.protocols#rpcv2Cbor",
+          // ... additional settings
+        }
+    }
+
+All protocols used for your service SHOULD use have an associated protocol trait applied to the service shape.
+For example to use the rpcv2 protocol with your service you would add the associated protocol trait to your Smithy model:
```

Comment:
A service declares compatible protocols by applying the corresponding protocol trait on the service shape. For example, if a service supports the rpcV2 protocol, it must have the protocol trait applied: